dominion/src/main/java/org/codice/dominion/Dominion.java

```diff
@@ -0,0 +1,175 @@
+
+/**
+ * The test runner to use to run with the Dominion framework.
+ *
+ * <p>This test runner actually delegates to a Dominion driver implementation which is itself a
+ * JUnit test runner.
+ *
+ * <p>Dominion discovers which driver to use by locating a registered {@link Factory} service on the
+ * classpath using Java's {@link ServiceLoader} class. It is possible to override this search by
+ * defining the system property {@link #DRIVER_FACTORY_PROPERTY} with the class name of a {@link
+ * Factory}.
+ *
+ * <p>Each driver should provide support for specification of JUnit method rules via the
+ * meta-annotation {@link org.codice.junit.ExtensionMethodRuleAnnotation}.
+ */
+public class Dominion extends Runner implements Filterable, Sortable {
+  private static final Logger LOGGER = LoggerFactory.getLogger(Dominion.class);
+
+  /** The type of driver this is. */
+  public static final String TYPE = "";
+
+  /** System property to use for overriding the driver search with a specific factory. */
+  public static final String DRIVER_FACTORY_PROPERTY = "dominion.driver.factory";
+
+  protected final Class<?> testClass;
+  protected final Runner delegate;
+  private final Filterable filterable;
+  private final Sortable sortable;
+
+  public Dominion(Class<?> testClass) {
+    this(testClass, Dominion.newDriver(testClass, ""));
+    LOGGER.debug("Dominion({})", testClass.getName());
+  }
+
+  protected Dominion(Class<?> testClass, Runner delegate) {
+    LOGGER.debug("Dominion({}, {})", testClass.getName(), delegate);
+    this.testClass = testClass;
+    this.delegate = delegate;
+    if (delegate instanceof Filterable) {
+      this.filterable = (Filterable) delegate;
+    } else {
+      this.filterable =
+          new Filterable() {
+            @Override
+            public void filter(Filter filter) throws NoTestsRemainException { // nothing to filter
+            }
+          };
+    }
+    if (delegate instanceof Sortable) {
+      this.sortable = (Sortable) delegate;
+    } else {
+      this.sortable = sorter -> {};
+    }
+  }
+
+  @Override
+  public Description getDescription() {
+    return delegate.getDescription();
+  }
+
+  @Override
+  public void run(RunNotifier notifier) {
+    delegate.run(notifier);
+  }
+
+  @Override
+  public void filter(Filter filter) throws NoTestsRemainException {
+    filterable.filter(filter);
+  }
+
+  @Override
+  public void sort(Sorter sorter) {
+    sortable.sort(sorter);
+  }
+
+  @Override
+  public String toString() {
+    return "Dominion[" + testClass.getName() + ", " + delegate + "]";
+  }
+
+  /**
+   * Factory which can be registered as a service retrievable via Java's {@link ServiceLoader} to
+   * let the Dominion framework know which driver classes to instantiate.
+   *
+   * <p><i>Note:</i> Defining the {@link #DRIVER_FACTORY_PROPERTY} system property will provide a
+   * factory which has priority over those registered via Java's {@link ServiceLoader}.
```

Comment:
Good question. 

A factory is an implied way to specify the driver to use. It would be use because you have that specific driver in your classpath. In some cases, you would have only one of these libraries in you classpath and as such, you are implying which driver to use.

If you have more than one of these driver libraries in your classpath, the selection would be based on the classpath search which is not always clear and it might be harder to know which one would be use. This is the downside of using Java SPI when you require only one implementation of a service.

The environment variable is a nice way to control from Maven for example that you want to force a particular driver. This might allow us to build blackbox test cases that are really independent of which type of container we use for running them (e.g. docker or pax-exam). In some case, it might be easier and faster to use PaxExam and probably easier to debug so we could specify to run the test cases with the PaxExam driver. In some other time, we might want to run some or all of those using docker. The environment variable provides us a way to override and force the driver we want.